File: java/ReorderList.java
/*ReorderList.java  
Reorder List  

Given a singly linked list L: L0 ->L1 ... ->Ln-1 ->Ln,
reorder it to: L0 ->Ln ->L1 ->Ln-1 ->L2 ->Ln-2 ->...

You must do this in-place without altering the nodes' values.

Example 

For example,
Given 1->2->3->4->null, reorder it to 1->4->2->3->null.

Tags Linked List 
*/

public class ReorderList {
    /**
     * @param head: The head of linked list.
     * @return: void
     */
    public static class ListNode {
       int val;
       ListNode next;
       ListNode(int val) {
           this.val = val;
           this.next = null;
       }
   	}

    public static ListNode reorderList(ListNode head) { 
        // write your code here
        if (head == null || head.next == null) {
            return head;
        }
        ListNode mid = findMid(head);
        //System.out.print(mid.val);
        ListNode right = reverse(mid.next);
        mid.next = null;
        ListNode left = head;
        ListNode result = merge(left,right);
        return result;
    }

    public static ListNode findMid(ListNode head) {
    	if (head == null || head.next == null) {
    		return head;
    	}
    	ListNode slow = head;
    	ListNode fast = head.next;
    	while (fast != null && fast.next != null) {
    		slow = slow.next;
    		fast = fast.next.next;
    	}
    	return slow;
    }

    public static ListNode merge(ListNode l1, ListNode l2) {
    	if (l1 == null) {
    		return l2;
    	}
    	if (l2 == null) {
    		return l1;
    	}

    	ListNode dummy = new ListNode(0);
    	ListNode cur = dummy;

    	while (l1 != null && l2 != null) {
    			cur.next = l1;
    			l1 = l1.next;
    			cur = cur.next;
    			cur.next = l2;
    			l2 = l2.next;
    			cur = cur.next;
    	}

    	if (l1 != null) {
    		cur.next = l1;
    	} else {
    		cur.next = l2;
    	}

    	return dummy.next;
    }

    public static ListNode reverse(ListNode head) {
     	if (head == null || head.next == null) {
     		return head;
     	}
        ListNode pre = null;
        while (head != null) {
	       	ListNode tmp = head.next;
	       	head.next = pre;
	       	pre = head;
	       	head = tmp;
        }
       return pre;
    }    
    public static void main(String[] args) {
        ReorderList test = new ReorderList();
        ListNode test1 = new ListNode(1);
        ListNode test2 = new ListNode(2);                                  
        ListNode test3 = new ListNode(3);
        ListNode test4 = new ListNode(4);
        ListNode test5 = new ListNode(5);
        ListNode test6 = new ListNode(6);
        ListNode test7 = new ListNode(7);
        ListNode test8 = new ListNode(8);
        ListNode test9 = new ListNode(9);

        test1.next = test2;
        test2.next = test3;
        test3.next = test4;
        test4.next = test5;
        test5.next = test6;
        test6.next = test7;
        test7.next = test8;
        test8.next = test9;
        test9.next = null;
        
        ListNode result = test.reorderList(test1);
        System.out.println("Expected: 1 9 2 8 3 7 4 6 5");
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;            
        }
    }
}

File: javaScript/reorderList.js
/*143. Reorder List  QuestionEditorial Solution  My Submissions
Total Accepted: 69010
Total Submissions: 295354
Difficulty: Medium
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
*/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
	//corner case
	if (head === null || head.next === null || head.next.next === null) {
		return head;
	}

	//divide
	var headHalf = new ListNode(-1);
	headHalf.next = head;
	var tailHalf = new ListNode(-1);

	var slow = head;
	var quick = head;

	while (quick !== null && quick.next !== null) {
		var pre_slow = slow;
		slow = slow.next;
		quick = quick.next.next;
	}

	if (quick.next === null) { // nodes number is even
		tailHalf.next = pre_slow.next;
		pre_slow.next = null;
	}

	if (quick === null) { // nodes number is odd
		tailHalf.next = slow.next;
		slow.next = null;
	}

	//reverse tailHalf
	var reversedTail = null;
	while( tailHalf !== null) {
		var tmp = tailHalf.next;
		tailHalf.next = reversedTail;
		reversedTail = tailHalf;
		tailHalf = tmp;
	}

	//merge
	var dummy = new ListNode(-1);
	var reorderHead = dummy;
	while (reversedTail !== null) {
		reorderHead.next = headHalf;
		reorderHead.next.next = reversedTail;
		headHalf = headHalf.next;
		reversedTail = reversedTail.next;
	}
	reorderHead.next = headHalf;
    
    return dummy.next;
};

//本题思路，用快慢指针把list分成前后两截，然后revserse后面一截，然后再merege: 结果OA显示： Do not return anything, modify head in-place instead.
//思考之后，修改如下：
var reorderList = function(head) {
	//corner case
	if (head === null || head.next === null || head.next.next === null) {
		return;
	}

	//divide: got tailhalf and set the end of headhalf
	var tailHalf = new ListNode(-1);

	var slow = head;
	var quick = head;

	while (quick !== null && quick.next !== null) {
		var pre_slow = slow;
		slow = slow.next;
		quick = quick.next.next;
	}

	if (quick === null) { // nodes number is even
		tailHalf.next = pre_slow.next;
		pre_slow.next = null;//set the end of headhalf
	} else if (quick.next === null) { // nodes number is odd
		tailHalf.next = slow.next;
		slow.next = null;//set the end of headhalf
	}

	//reverse tailHalf
	var reversedTail = null;
	while( tailHalf!== null) {
		var tmp = tailHalf.next;
		tailHalf.next = reversedTail;
		reversedTail = tailHalf;
		tailHalf = tmp;
	}

	//insert reversed tail half
	var runner = head;
	while (runner !== null && reversedTail.next !== null) { // stop before the -1 dummy head
		var temp = reversedTail.next;
		reversedTail.next = runner.next;
		runner.next = reversedTail;
		reversedTail = temp;
		if (runner.next !== null) {
			runner = runner.next.next			
		} else {
			runner = runner.next;
		}
	}
    
    return;
};